crates/pairwise_nostd_internal/src/misc.rs

```diff
@@ -94,6 +94,123 @@ pub fn segment_idx_bounds(n_indices: usize, seg_index: usize, n_segments: usize)
+    // - we should consider moving away from using z, y, and x since those can
+    //   be mapped arbitrarily. Instead I think we should talk about axis-0,
+    //   axis-1, and axis-2 (the way numpy does it)
+    shape_zyx: [isize; 3],
+    strides_zyx: [isize; 3],
+}
+
+#[allow(dead_code)] // <- this is a temporary stopgap solution
+impl View3DProps {
+    /// Create a contiguous-in-memory View3DProps from shape_zyx alone
+    pub fn from_shape_contiguous(shape_zyx: [usize; 3]) -> Result<View3DProps, &'static str> {
+        check_shape(&shape_zyx)?;
+        Ok(Self {
+            shape_zyx: [
+                shape_zyx[0] as isize,
+                shape_zyx[1] as isize,
+                shape_zyx[2] as isize,
+            ],
+            strides_zyx: [
+                (shape_zyx[1] * shape_zyx[2]) as isize,
+                shape_zyx[2] as isize,
+                1_isize,
+            ],
+        })
+    }
+
+    /// Create a View3DProps from shape_zyx and strides_zyx
+    pub fn from_shape_strides(
+        shape_zyx: [usize; 3],
+        strides_zyx: [usize; 3],
+    ) -> Result<View3DProps, &'static str> {
+        check_shape(&shape_zyx)?;
+
+        if strides_zyx[2] != 1 {
+            Err("the blocks must be contiguous along the fast axis")
+        } else if strides_zyx[1] < shape_zyx[2] * strides_zyx[2] {
+            Err("the length of the contiguous axis can't exceed strides_zyx[1]")
+        } else if strides_zyx[0] < shape_zyx[1] * strides_zyx[1] {
+            Err("the length of axis 1 can't exceed strides_zyx[0]")
+        } else if strides_zyx[0] < strides_zyx[1] {
+            Err("strides_zyx[1] must not exceed strides_zyx[0]")
+        } else {
+            Ok(Self {
+                shape_zyx: [
+                    shape_zyx[0] as isize,
+                    shape_zyx[1] as isize,
+                    shape_zyx[2] as isize,
+                ],
+                strides_zyx: [
+                    strides_zyx[0] as isize,
+                    strides_zyx[1] as isize,
+                    strides_zyx[2] as isize,
+                ],
+            })
+        }
+    }
+
+    /// returns the number of elements that a slice must have to be described
+    /// by self
+    pub fn contiguous_length(&self) -> usize {
+        let max_idx_1d = self.map_idx(
+            self.shape_zyx[0] - 1,
+            self.shape_zyx[1] - 1,
+            self.shape_zyx[2] - 1,
+        );
+        // based on the invariants enforced by the constructor, tmp is always
+        // always be positive. Thus, it can be losslessly converted to usize
+        (max_idx_1d as usize) + 1_usize
+    }
+
+    pub fn shape(&self) -> &[isize; 3] {
+        &self.shape_zyx
+    }
+
+    /// map a 3D index to 1D
+    pub fn map_idx(&self, iz: isize, iy: isize, ix: isize) -> isize {
+        iz * self.strides_zyx[0] + iy * self.strides_zyx[1] + ix
+    }
+
+    /// map a 1D index to a 3D index
+    pub fn reverse_map_idx(&self, idx: isize) -> [isize; 3] {
```

Comment:
I don't love this name, but I'm having trouble thinking of something better.  `invert_idx`? `dimensional_idx`?